File: frontend/src/types/events.ts
export type SSEEventType = 'run.start' | 'text.delta' | 'run.finish' | 'conversation.title' | 'tool.call' | 'tool.result' | 'messages.sent' | 'provider.fallback' | 'error'

export interface RunStartData {
  run_id: string
  provider: string
  model: string
}

export interface TextDeltaData {
  content: string
}

export interface RunFinishData {
  finish_reason: string
  token_usage?: { prompt_tokens: number; completion_tokens: number; total_tokens: number }
}

export interface ConversationTitleData {
  title: string
}

export interface ToolCallData {
  name: string
  arguments: Record<string, unknown>
  step_index?: number
}

export interface ToolResultData {
  name: string
  result: Record<string, unknown>
  step_index?: number
}

export interface ProviderFallbackData {
  from_provider: string
  to_provider: string
  step_index: number
}

export interface MessagesSentData {
  call_index: number
  messages: Array<{ role: string; content: string }>
}

export interface ErrorData {
  message: string
}

export interface SSEEvent {
  type: SSEEventType
  ts: string
  data: RunStartData | TextDeltaData | RunFinishData | ConversationTitleData | ToolCallData | ToolResultData | MessagesSentData | ProviderFallbackData | ErrorData
}

File: backend/src/agent_chat/schemas/events.py
"""SSE event type definitions."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class RunStartData(BaseModel):
    run_id: str
    provider: str
    model: str


class TextDeltaData(BaseModel):
    content: str


class RunFinishData(BaseModel):
    finish_reason: str
    token_usage: dict | None = None


class ConversationTitleData(BaseModel):
    title: str


class ToolCallData(BaseModel):
    tool_name: str
    arguments: dict
    risk_level: str = "read"


class ToolResultData(BaseModel):
    tool_name: str
    result: dict
    error_code: str | None = None


class ErrorData(BaseModel):
    message: str


class SSEEvent(BaseModel):
    type: str
    ts: datetime
    data: dict
